src/query/state.rs

```diff
@@ -66,6 +75,48 @@ impl QueryState {
     }
 }
 
+pub struct RunningQuery {
+    pub result: Receiver<QueryResult>,
+
+    /// `JoinHandle` for the query task.
+    ///
+    /// The join handle is only useful for the purpose of aborting the query. Tasks started with
+    /// `tokio::spawn` run to completion whether or not anything waits on the handle.
+    ///
+    /// We could return the result via the JoinHandle, except that we want to check the status
+    /// of the task, and shuttle doesn't implement `JoinHandle::is_finished`.
+    pub join_handle: JoinHandle<()>,
+}
+
+impl RunningQuery {
+    pub fn try_complete(&mut self) -> Option<QueryResult> {
+        match self.result.try_recv() {
+            Ok(result) => Some(result),
+            Err(TryRecvError::Closed) => {
```

Comment:
wouldn't it also be in case if `try_complete` is called after the results have been received already? Would a lingering HTTP request that arrives right in between "try_complete".."complete" sequence of calls crash our query?

Reply:
I think we are safe here because `RunningQuery` is behind a mutex, so it is not possible to poll it from two different threads

Reply:
Yeah, I think we're okay because of the mutex. Note that `try_complete` and `try_recv` take `&mut self`, so even if something were to change about the mutex, the mutability rules provide some protection.

I think in the long term we'll need some kind of "query manager" component that is free-running in the server. The query manager would move a query from running to completed state soon after it finishes, rather than the query remaining in the running state until somebody happens to call the status API. The query manager would also be responsible for things like garbage collecting abandoned queries after some period of time. In the case where there are multiple things asking for query status (e.g. the client APIs and a management API), it will not make sense for the client status API to be responsible for bookkeeping query status.

I didn't intend `RunningQuery` to be anything complicated -- I called it that because it's the data associated with the `QueryState::Running` enum variant. If we get `JoinHandle::is_finished` in shuttle, then we can get rid of `RunningQuery` and go back to `QueryState::Running(JoinHandle)`. It seems redundant to me to put the QueryId in RunningQuery -- I would rather use `QueryHandle` (which already contains QueryId) for that, possibly augmented with a method or `Deref` to access `QueryState`. (It also seems redundant to me for the AwaitingInputs state to contain QueryId.)